quickwit-serve/src/search_api/rest_handler.rs

```diff
@@ -166,44 +166,59 @@ fn search_post_filter(
         .and(warp::body::json())
 }
 
-async fn search<TSearchService: SearchService>(
+async fn search(
     index_id: String,
     search_request: SearchRequestQueryString,
-    search_service: Arc<TSearchService>,
+    search_service: Arc<dyn SearchService>,
 ) -> Result<impl warp::Reply, Infallible> {
     info!(index_id = %index_id, request =? search_request, "search");
     Ok(search_request
         .format
         .make_reply(search_endpoint(index_id, search_request, &*search_service).await))
 }
 
+fn require<T: Clone + Send>(
+    val_opt: Option<T>,
+) -> impl Filter<Extract = (T,), Error = Rejection> + Clone {
+    warp::any().and_then(move || {
+        let val_opt_clone = val_opt.clone();
+        async move {
+            if let Some(val) = val_opt_clone {
+                Ok(val)
+            } else {
+                Err(warp::reject())
+            }
+        }
+    })
+}
+
 /// REST GET search handler.
 ///
 /// Parses the search request from the
-pub fn search_get_handler<TSearchService: SearchService>(
-    search_service: Arc<TSearchService>,
+pub fn search_get_handler(
+    search_service_opt: Option<Arc<dyn SearchService>>,
 ) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
     search_get_filter()
-        .and(warp::any().map(move || search_service.clone()))
+        .and(require(search_service_opt))
```

Comment:
nice! I did not think about using this trick to always return same impl.

I was thinking about adding/removing handlers depending on where the service is active. Of course, there is one small cons: every hander must be something like `BoxFilter<dyn Replay>`.

Reply:
correct.